### PR TITLE
Fix #2021: Seperate robolectric check provider

### LINKED ISSUE #2021: Separate out Robolectric Check Provider
**Is your feature request related to a problem? Please describe.**
Currently, the  Robolectric Check Provider is present in the `TestDispatcherModule`. 

**Describe the solution you'd like**
Separate it out. Creating a different provider for it, as this is something which can use commonly in other modules/tests also.

### PR DESCRIPTION
**Pull request by FareesHussain**
## Explanation

Fixes #2021

 - [Throughly tested](https://github.com/FareesHussain/oppia-android/actions?query=branch%3Aseperate-robolectric-check-provider) 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt
Comment (anandwana001) [inline data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/audio/AudioPlayerControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/audio/AudioPlayerControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/audio/CellularAudioDialogControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/audio/CellularAudioDialogControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/AnalyticsControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/AnalyticsControllerTest.kt]: Do we need this here?

File: domain/src/test/java/org/oppia/android/domain/oppialogger/exceptions/ExceptionsControllerTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/oppialogger/exceptions/ExceptionsControllerTest.kt]: Do we need this here? and other non required places!!

File: testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt
Comment (anandwana001) [inline testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]: need kdoc for this file.

File: testing/src/main/java/org/oppia/android/testing/TestDispatcherModule.kt
Comment (anandwana001) [inline testing/src/main/java/org/oppia/android/testing/TestDispatcherModule.kt:88]: Why are we not removing this completely from here?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@FareesHussain

File: data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt
Comment (FareesHussain) [reply data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt]: @anandwana001 TestDispatcherModule.kt contains provideTestCoroutineDispatchers and provideTestCoroutineDispatcherFactory that require `@IsOnRobolectric` annotation as a param 

so that might be the reason why we need to add RobolectricProviderModule for all Test files in which TestDispatcherModule is used

And there are very few test files that only require RobolectricProviderModule

File: testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt
Comment (FareesHussain) [reply testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]:
```
/**
 * Dagger [Module] that provides [IsOnRobolectric]
 */
```
I've added this comment for now does this look good

File: testing/src/main/java/org/oppia/android/testing/TestDispatcherModule.kt
Comment (FareesHussain) [reply testing/src/main/java/org/oppia/android/testing/TestDispatcherModule.kt:88]: I forgot to remove this completely I'll push this in a commit with above kdoc you suggest

File: data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt
Comment (anandwana001) [reply data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt]: ah, ok. Is this separation causing any test fails?

Comment (FareesHussain) [reply data/src/test/java/org/oppia/android/data/persistence/PersistentCacheStoreTest.kt]: no

File: testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt
Comment (anandwana001) [inline testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]: no empty line

File: testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt
Comment (anandwana001) [inline testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]: Can we be more clear what does this module provide and when it should be used.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@FareesHussain Left a comment.
Name sounds a bit odd, provider module. 
WDYT about `RobolectricModule.kt`, we can use it as if in future we need anything related to Robolectric from this module.

File: testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt
Comment (FareesHussain) [reply testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]: @anandwana001 I'm not sure where this is used can you please help me with updating the kdocs for this
also we need to update the kdocs for TestDispatcherModule it depends on RobolectricModule right ?

Comment (anandwana001) [reply testing/src/main/java/org/oppia/android/testing/RobolectricProviderModule.kt]: not used, this doc should be about when to use this file, so it helps any developer to understand what are the cases when this module is usable.

**Comment by FareesHussain** [comment]
The test case `testWorker_logException_withoutNetwork_enqueueRequest_verifySuccess` from LogUploadWorkerTest.kt 
 fails most of the time 
image
can this be fixed

**Comment by BenHenning** [comment]
Hi! Sorry, will need to take a look at this tomorrow my time.

**Comment by anandwana001** [comment]
> The test case `testWorker_logException_withoutNetwork_enqueueRequest_verifySuccess` from LogUploadWorkerTest.kt
> fails most of the time
> image
> can this be fixed

This seems like a flaky test, most of the time it passed on actions check and sometimes not.

**Review by BenHenning — APPROVED** [review]
Thanks @FareesHussain! Overall, I think this looks good. This PR is another reminder of how much we'd benefit from switching over to Dagger Hilt, as well. :) (this PR would have only been ~2 files large in the Hilt world).

**Review by anandwana001 — APPROVED** [review]
LGTM

**Comment by BenHenning** [comment]
@rt4914 I think you need to do a codeowners review of this.

**Comment by FareesHussain** [comment]
@rt4914 Please don't merge this pr and #2233 at the same time,
as it involves creating a new test Suite and I need to inject the RobolectricModule

**Comment by rt4914** [comment]
> @rt4914 Please don't merge this pr and #2233 at the same time,
> as it involves creating a new test Suite and I need to inject the RobolectricModule

In that case, can you update this PR so that we can merge this within hour and later on we can focus on other PR? @FareesHussain

**Comment by FareesHussain** [comment]
@rt4914 This can be merged now if this is merged I'll just update the other pr

**Comment by anandwana001** [comment]
@FareesHussain please check the conflicts.

**Comment by FareesHussain** [comment]
> @FareesHussain please check the conflicts.

Fixed @rt4914 please check

**Review by rt4914 — APPROVED** [review]
LGTM, thanks. Merging now.

**Review by vinitamurthi — APPROVED** [review]

**Pull request merged**
